lapack/testlapack/dlag2.go

```diff
@@ -0,0 +1,96 @@
+	// The scaling factor "s" results in a modified eigenvalue equation
+	//  s A - w B
+	// where  s  is a non-negative scaling factor chosen so that  w,  w B,
+	// and  s A  do not overflow and, if possible, do not underflow, either.
+	//
+	// B is a  ldb, 2)
+	// On entry, the 2 x 2 upper triangular matrix B.  It is
+	// assumed that the one-norm of B is less than 1/SAFMIN.  The
+	// diagonals should be at least sqrt(SAFMIN) times the largest
+	// element of B (in absolute value); if a diagonal is smaller
+	// than that, then  +/- sqrt(SAFMIN) will be used instead of
+	// that diagonal.
+	//
+	// Dlag2 is an internal routine. It is exported for testing purposes.
+	Dlag2(a []float64, lda int, b []float64, ldb int) (scale1, scale2, wr1, wr2, wi float64)
+}
+
+func Dlag2Test(t *testing.T, impl Dlag2er) {
+	rnd := rand.New(rand.NewSource(1))
+	for _, lda := range []int{2, 3, 4, 5, 6, 7, 30} {
+		for _, ldb := range []int{2, 3, 4, 5, 6, 7, 30} {
+			dlag2Test(t, impl, rnd, lda, ldb)
+		}
+	}
+}
+
+func dlag2Test(t *testing.T, impl Dlag2er, rnd *rand.Rand, lda, ldb int) {
+	const tol = 1e-14
+	a := randomGeneral(2, 2, lda, rnd)
+	b := randomGeneral(2, 2, ldb, rnd)
+	b.Data[ldb] = 0 // Dlag2 requires b to be upper triangular
+	a11 := a.Data[0]
+	a12 := a.Data[1]
+	a21 := a.Data[lda]
+	a22 := a.Data[lda+1]
+	b11 := b.Data[0]
+	b12 := b.Data[1]
+	b22 := b.Data[ldb+1]
+	name := fmt.Sprintf("a11=%.4v,a12=%.4v,a21=%.4v,a22=%.4v,b11=%.4v,b12=%.4v,b22=%.4v", a11, a12, a21, a22, b11, b12, b22)
+	scale1, scale2, wr1, wr2, wi := impl.Dlag2(a.Data, a.Stride, b.Data, b.Stride)
+	if wi < 0 {
+		t.Errorf("%v: wi can not be negative wi=%v", name, wi)
+	}
+	// Complex eigenvalue solution.
+	if wi > 0 {
+		if wr1 != wr2 {
+			t.Errorf("%v: wr1 should equal wr2 for complex eigenvalues. got %v!=%v", name, wr1, wr2)
+		}
+		idet1 := cmplxdet2x2(complex(scale1*a11-wr1*b11, -wi*b11), complex(scale1*a12-wr1*b12, -wi*b12),
+			complex(scale1*a21, 0), complex(scale1*a22-wr1*b22, -wi*b22))
+		idet2 := cmplxdet2x2(complex(scale1*a11-wr1*b11, wi*b11), complex(scale1*a12-wr1*b12, wi*b12),
+			complex(scale1*a21, 0), complex(scale1*a22-wr1*b22, wi*b22))
+
+		if math.Abs(real(idet1))+math.Abs(imag(idet1)) > tol || math.Abs(real(idet2))+math.Abs(imag(idet2)) > tol {
+			t.Errorf("%v: (%.4v±%.4vi)/%.4v did not solve eigenvalue problem", name, wr1, wi, scale1)
+		}
+		return // Complex solution verified
+	}
+
+	// Real eigenvalue solution.
+	// |s A - w B| = 0   is solution for eigenvalue problem. Calculate distance from solution.
+	res1 := det2x2(scale1*a11-wr1*b11, scale1*a12-wr1*b12,
+		scale1*a21, scale1*a22-wr1*b22)
+	if math.Abs(res1) > tol {
+		t.Errorf("%v: got a residual from |%0.4v*A - w1*B| > tol: %v", name, scale1, res1)
+	}
+	res2 := det2x2(scale2*a11-wr2*b11, scale2*a12-wr2*b12,
+		scale2*a21, scale2*a22-wr2*b22)
+	if math.Abs(res2) > tol {
+		t.Errorf("%v: got a residual from |s*A - w2*B| > tol: %v", name, res1)
+	}
+}
+
+// Solves complex determinant of
```

Comment:
```suggestion
// cmplxdet2x2 returns the determinant of
```
and please the body of the function on its own line.